specs/test_formats/README.md

```diff
@@ -0,0 +1,170 @@
+  - A generator should only output one `type` of test.
+  - A generator is free to output multiple `suite` files, optionally with different `handler`s.
+- `type`: the specialization of one single `generator`.
+- `suite`: a YAML file with:
+  - a header: describes the `suite`, and defines what the `suite` is for
+  - a list of test cases
+- `runner`: where a generator is a *"producer"*, this is the *"consumer"**.
+  - A `runner` focuses on *only one* `type`, and each type has *only one* `runner`.
+- `handler`: a `runner` may be too limited sometimes, you may have a `suite` with a specific focus that requires a different format.
+  To facilitate this, you specify a `handler`: the runner can deal with the format by using the specified handler.
+  Using a `handler` in a `runner` is optional.
+- `case`: a test case, an entry in the `test_cases` list of a `suite`. A case can be anything in general, 
+  but its format should be well-defined in the documentation corresponding to the `type` (and `handler`).\
+  A test has the same exact configuration and fork context as the other entries in the `case` list of its `suite`.
+- `forks_timeline`: a fork timeline definition, a YAML file containing a key for each fork-name, and a slot number as value.
+
+## Test format philosophy
+
+### Config design
+
+After long discussion, the following types of configured constants were identified:
+- Never changing: genesis data
+- Changing, but reliant on old value: e.g. a slot time may change, but if you want to do the conversion 
+  `(genesis data, timestamp) -> slot number` you end up needing both constants.
+- Changing, but kept around during fork transition: finalization may take a while,
+  e.g. an executable has to deal with new deposits and old deposits at the same time. Another example may be economic constants.
+- Additional, back-wards compatible: new constants are introduced for later phases
+- Changing: there is a very small chance some constant may really be *replaced*. 
+  In this off-chance, it is likely better to include it as an additional variable,
+  and some clients may simply stop supporting the old one, if they do not want to sync from genesis.
+
+Based on these types of changes, we model the config as a list of key value pairs,
+ that only grows with every fork (they may change in development versions of forks however, git manages this).
+With this approach, configurations are backwards compatible (older clients ignore unknown variables), and easy to maintain.
+
+### Fork config design
+
+There are two types of fork-data:
+1) timeline: when does a fork take place?
+2) coverage: what forks are covered by a test?
+
+The first is neat to have as a separate form: we prevent duplication, and can run with different presets
+ (e.g. fork timeline for a minimal local test, for a public testnet, or for main-net)
+
+The second is still somewhat ambiguous: some tests may want cover multiple forks, and can do so in different ways:
```

Comment:
IMO it's unlikely that we *need* to have one test suite for multiple forks. Even the same scenario of Petersburg and Byzantium happen (fork on the same block) in eth2, the test suite only need to provide  Byzantium tests.

Reply:
The idea is that for some tests, they should be exactly the same for some consecutive forks. And instead of duplicating the file to ensure it runs as expected for multiple forks, I would use the forks-list for these types of tests to indicate this consistency explicitly. And, it adds to another property: we can really easily determine exactly which forks are covered by which tests.

Reply:
the 1.0 tests actually do a version of this by specifying the results of a test for multiple fork versions within the same file and leave them blank for unsupported forks. Very often this just ends up with overly verbose test files just repeating the same outputs IMO. (example here https://github.com/ethereum/tests/blob/develop/TransactionTests/ttVValue/V_equals38.json)

This instead is a list to make explicit which forks a test is intended to run against. So if it were an SSZ test for a feature that was supported in three forks, the three forks would be noted. If the feature changed in a forth fork, the forth fork would not be added to the list and instead a new tests would be written noting the forth fork.

Reply:
Ahh sorry I misunderstood the "cover" here; I mixed up it with the timeline thing and thought you want to make one test case includes multiple state transition**s** with different forks. 😅 

Now I understand it's about compatibility and yes that makes sense to me. 😃 👍 👍 
FYI in eth1, they also use `>=` notation for compatibility like:
https://github.com/ethereum/tests/blob/5cef36e2844d4cfdc6b52362ba09261f0a09b76e/src/GeneralStateTestsFiller/stRevertTest/TouchToEmptyAccountRevert3Filler.json#L18

Reply:
With conditions it's somewhat harder to parse. I'm open for it, but just feel like a list works well enough. And we won't have any ambiguity when there are two forks at the same time, and you need to decide what to include/exclude.

Reply:
Fixed a bit of wording, and included example configs for fork-timelines and constant presets, with format description.